### ISSUE TITLE
analysis server crash: tryin calling isSubtypeOf on null – PostfixCompletionProcessor._findOuterExpression

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#34114]
Analyzer Feedback from IntelliJ

# Version information

- `IDEA IU-182.3911.36`
- `2.1.0-dev.0.0`
- `IU-182.3911.36, JRE 1.8.0_152-release-1248-b8x64 JetBrains s.r.o, OS Mac OS X(x86_64) v10.13.6 unknown, screens 3840x2160`

# Exception

```
Dart analysis server, SDK version 2.1.0-dev.0.0, server version 1.20.5, FATAL error: Failed to handle request: {id: 823, method: edit.isPostfixCompletionApplicable, params: {file: /Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart, offset: 3046, key: .return}, clientRequestTime: 1533853312500}
NoSuchMethodError: The method 'isSubtypeOf' was called on null.
Receiver: null
Tried calling: isSubtypeOf(Instance of 'InterfaceTypeImpl')
#0      Object.noSuchMethod (dart:core/runtime/libobject_patch.dart:47:5)
#1      PostfixCompletionProcessor._findOuterExpression.<anonymous closure> (package:analysis_server/src/services/completion/postfix/postfix_completion.dart:566:16)
#2      _ListBase&Object&ListMixin.firstWhere (dart:collection/list.dart:142:15)
#3      PostfixCompletionProcessor._findOuterExpression (package:analysis_server/src/services/completion/postfix/postfix_completion.dart:564:28)
#4      PostfixCompletionProcessor.findObjectExpression (package:analysis_server/src/services/completion/postfix/postfix_completion.dart:482:7)
#5      DartPostfixCompletion.isObjectContext (package:analysis_server/src/services/completion/postfix/postfix_completion.dart:221:22)
#6      PostfixCompletionProcessor.isApplicable (package:analysis_server/src/services/completion/postfix/postfix_completion.dart:513:23)
<asynchronous suspension>
#7      EditDomainHandler.isPostfixCompletionApplicable (package:analysis_server/src/edit/edit_domain.dart:443:33)
<asynchronous suspension>
#8      EditDomainHandler.handleRequest (package:analysis_server/src/edit/edit_domain.dart:364:9)
#9      AnalysisServer.handleRequest.<anonymous closure>.<anonymous closure> (package:analysis_server/src/analysis_server.dart:744:45)
#10     _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:210:15)
#11     AnalysisServer.handleRequest.<anonymous closure> (package:analysis_server/src/analysis_server.dart:740:50)
#12     _rootRun (dart:async/zone.dart:1124:13)
#13     _CustomZone.run (dart:async/zone.dart:1021:19)
#14     _runZoned (dart:async/zone.dart:1516:10)
#15     runZoned (dart:async/zone.dart:1500:12)
#16     AnalysisServer.handleRequest (package:analysis_server/src/analysis_server.dart:739:5)
#17     ByteStreamServerChannel._readRequest.<anonymous closure> (package:analysis_server/src/channel/byte_stream_channel.dart:174:16)
#18     _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:210:15)
#19     ByteStreamServerChannel._readRequest (package:analysis_server/src/channel/byte_stream_channel.dart:165:47)
#20     ByteStreamServerChannel.listen.<anonymous closure> (package:analysis_server/src/channel/byte_stream_channel.dart:110:26)
#21     _rootRunUnary (dart:async/zone.dart:1132:38)
#22     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#23     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931:7)
#24     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:336:11)
#25     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:263:7)
#26     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:68:11)
#27     _EventSinkWrapper.add (dart:async/stream_transformers.dart:15:11)
#28     _StringAdapterSink.add (dart:convert/string_conversion.dart:268:11)
#29     _LineSplitterSink._addLines (dart:convert/line_splitter.dart:154:13)
#30     _LineSplitterSink.addSlice (dart:convert/line_splitter.dart:129:5)
#31     StringConversionSinkMixin.add (dart:convert/string_conversion.dart:189:5)
#32     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:120:24)
#33     _rootRunUnary (dart:async/zone.dart:1132:38)
#34     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#35     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931:7)
#36     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:336:11)
#37     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:263:7)
#38     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:68:11)
#39     _EventSinkWrapper.add (dart:async/stream_transformers.dart:15:11)
#40     _StringAdapterSink.add (dart:convert/string_conversion.dart:268:11)
#41     _StringAdapterSink.addSlice (dart:convert/string_conversion.dart:273:7)
#42     _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:348:20)
#43     _Utf8ConversionSink.add (dart:convert/string_conversion.dart:341:5)
#44     _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:86:18)
#45     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:120:24)
#46     _rootRunUnary (dart:async/zone.dart:1132:38)
#47     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#48     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931:7)
#49     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:336:11)
#50     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:263:7)
#51     _SyncStreamController._sendData (dart:async/stream_controller.dart:763:19)
#52     _StreamController._add (dart:async/stream_controller.dart:639:7)
#53     _StreamController.add (dart:async/stream_controller.dart:585:5)
#54     _Socket._onData (dart:io/runtime/binsocket_patch.dart:1714:41)
#55     _rootRunUnary (dart:async/zone.dart:1136:13)
#56     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#57     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931:7)
#58     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:336:11)
#59     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:263:7)
#60     _SyncStreamController._sendData (dart:async/stream_controller.dart:763:19)
#61     _StreamController._add (dart:async/stream_controller.dart:639:7)
#62     _StreamController.add (dart:async/stream_controller.dart:585:5)
#63     new _RawSocket.<anonymous closure> (dart:io/runtime/binsocket_patch.dart:1276:33)
#64     _NativeSocket.issueReadEvent.issue (dart:io/runtime/binsocket_patch.dart:819:14)
#65     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#66     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#67     _runPendingImmediateCallback (dart:isolate/runtime/libisolate_patch.dart:114:13)
#68     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:170:5)
```

```
1533853311802 <= {"event":"completion.results","params":{"id":"231","replacementOffset":3040,"replacementLength":0,"results":[],"isLast":true}}
1533853312102 => {"id":"818","method":"edit.getFixes","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","offset":3035},"clientRequestTime"
1533853312103 <= {"id":"818","result":{"fixes":[]}}
1533853312103 => {"id":"819","method":"analysis.updateContent","params":{"files":{"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart":{"type":"add","content"
1533853312105 <= {"id":"819","result":{}}
1533853312106 <= {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
1533853312111 => {"id":"820","method":"edit.getAssists","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","offset":3046,"length":0},"clien
1533853312119 <= {"event":"analysis.errors","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","errors":[]}}
1533853312120 <= {"id":"820","result":{"assists":[{"message":"Convert to block documentation comment","edits":[{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_seria
1533853312121 <= {"event":"analysis.highlights","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","regions":[{"type":"KEYWORD","offset":29
1533853312122 <= {"event":"analysis.navigation","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","regions":[{"offset":275,"length":16,"ta
1533853312122 <= {"event":"analysis.closingLabels","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","labels":[]}}
1533853312124 <= {"event":"analysis.outline","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","kind":"LIBRARY","outline":{"element":{"kin
1533853312124 <= {"event":"analysis.overrides","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","overrides":[]}}
1533853312126 <= {"event":"server.status","params":{"analysis":{"isAnalyzing":false}}}
1533853312132 <= {"event":"analysis.implemented","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","classes":[{"offset":3767,"length":7}],
1533853312486 => {"id":"821","method":"analysis.updateContent","params":{"files":{"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart":{"type":"add","content"
1533853312486 => {"id":"822","method":"completion.getSuggestions","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","offset":3047},"client
1533853312487 <= {"id":"821","result":{}}
1533853312487 <= {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
1533853312496 <= {"id":"822","result":{"id":"232"}}
1533853312496 <= {"event":"analysis.errors","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","errors":[]}}
1533853312497 <= {"event":"completion.results","params":{"id":"232","replacementOffset":3047,"replacementLength":0,"results":[],"isLast":true}}
1533853312498 <= {"event":"analysis.highlights","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","regions":[{"type":"KEYWORD","offset":29
1533853312498 <= {"event":"analysis.navigation","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","regions":[{"offset":275,"length":16,"ta
1533853312499 <= {"event":"analysis.closingLabels","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","labels":[]}}
1533853312499 <= {"event":"analysis.outline","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","kind":"LIBRARY","outline":{"element":{"kin
1533853312499 <= {"event":"analysis.overrides","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","overrides":[]}}
1533853312500 => {"id":"823","method":"edit.isPostfixCompletionApplicable","params":{"file":"/Users/kevmoo/source/github/json_serializable/json_annotation/lib/src/json_serializable.dart","offset":3046
1533853312511 <= {"event":"server.error","params":{"isFatal":true,"message":"Failed to handle request: {id: 823, method: edit.isPostfixCompletionApplicable, params: {file: /Users/kevmoo/source/github/
```

**Comment by bwilkerson** [comment]
@stevemessick Could you take a look at this? The code is assuming that every expression will have a static type, which isn't true in the face of errors, but I'm not familiar enough with the code to put together a test case to reproduce the exception.

**Comment by devoncarew** [comment]
@stevemessick, we're seeing regular reports of this - can you take a look? I've hit it myself a few times, went to report the issue, and saw several more occurrences in the issue tracker.

**Comment by kevmoo** [comment]
I hit this daily!

**Comment by kevmoo** [comment]
Again with `2.1.0-dev.3.0`

**Comment by scheglov** [comment]
Fixed as 83324a8db93cb2bea6b7a0720d6b760656763efb
